test/src/test_hetd2.cpp

```diff
@@ -0,0 +1,150 @@
+/// @file test_hetd2.cpp
+/// @author Skylar Johns, University of Colorado Denver, USA
+/// @author Weslley S Pereira, University of Colorado Denver, USA
+/// @brief Test HETD2
+//
+// Copyright (c) 2021-2023, University of Colorado Denver. All rights reserved.
+//
+// This file is part of <T>LAPACK.
+// <T>LAPACK is free software: you can redistribute it and/or modify it under
+// the terms of the BSD 3-Clause license. See the accompanying LICENSE file.
+
+#include <catch2/catch_template_test_macros.hpp>
+#include <catch2/generators/catch_generators.hpp>
+
+// Test utilities and definitions (must come before <T>LAPACK headers)
+#include "testutils.hpp"
+
+// Auxiliary routines
+#include <tlapack/lapack/lacpy.hpp>
+#include <tlapack/lapack/lange.hpp>
+// #include <tlapack/plugins/debugutils.hpp>
+
+// Other routines
+#include <tlapack/blas/gemm.hpp>
+#include <tlapack/lapack/hetd2.hpp>
+#include <tlapack/lapack/ungtr.hpp>
+
+using namespace tlapack;
+
+TEMPLATE_TEST_CASE("Tridiagnolization of a symmetric matrix works",
+                   "[hetd2]",
+                   TLAPACK_TYPES_TO_TEST)
+{
+    srand(1);
+
+    using matrix_t = TestType;
+    using T = type_t<matrix_t>;
+    using idx_t = size_type<matrix_t>;
+    typedef real_type<T> real_t;
+
+    // Functor
+    Create<matrix_t> new_matrix;
+
+    // Generators
+    idx_t n = GENERATE(6, 13, 29);
+    const Uplo uplo = GENERATE(Uplo::Lower, Uplo::Upper);
+
+    DYNAMIC_SECTION("n = " << n << " uplo = " << uplo)
+    {
+        // Constants
+        const real_t zero(0);
+        const real_t one(1);
+        const real_t eps = ulp<real_t>();
+        const real_t tol = real_t(2 * n) * eps;
+
+        // Matrices and vectors
+        std::vector<T> A_;
+        auto A = new_matrix(A_, n, n);
+        std::vector<T> Q_;
+        auto Q = new_matrix(Q_, n, n);
+        std::vector<real_t> E(n - 1), D(n);
+        std::vector<T> tau(n - 1);
+
+        // Fill A with random values
+        for (idx_t j = 0; j < n; ++j)
+            for (idx_t i = 0; i < n; ++i)
+                A(i, j) = rand_helper<T>();
```

Comment:
perhaps you can fill the parts of A that should not be accessed with NaN or something and then later check that they were not changed.

Reply:
Thanks! That is a good idea.